File: android/app/src/main/java/dev/pepotech/pepomote/ui/components/CollapsibleHeader.kt
package dev.pepotech.pepomote.ui.components

import android.content.Context
import android.view.accessibility.AccessibilityManager
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.tween
import androidx.compose.animation.expandVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.pointer.PointerEventPass
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import dev.pepotech.pepomote.R
import dev.pepotech.pepomote.service.LinkState
import dev.pepotech.pepomote.service.UiLink
import dev.pepotech.pepomote.ui.screens.ModeChips
import dev.pepotech.pepomote.ui.screens.NunchukChip
import dev.pepotech.pepomote.ui.screens.isWiiUAsWiimote
import dev.pepotech.pepomote.ui.screens.modeLabel
import dev.pepotech.pepomote.ui.screens.showModeChips
import dev.pepotech.pepomote.ui.screens.showNunchukChip
import dev.pepotech.pepomote.ui.theme.PepoColors
import kotlinx.coroutines.delay

/**
 * Cuándo se pliega y cuándo se despliega sola la cabecera de los mandos
 * apaisados. Sin Compose: se prueba en la JVM.
 */
object HeaderCollapse {
    /** Lo que tarda en plegarse sola tras el último toque. */
    const val AUTO_MS = 4000L

    /**
     * Se pliega sola solo con el enlace vivo (sin él hace falta «Salir» a la
     * vista) y nunca con un lector de pantalla: ahí el usuario va tocando y
     * escuchando, y una cabecera que desaparece sola se pierde.
     */
    fun autoCollapses(connected: Boolean, screenReader: Boolean): Boolean = connected && !screenReader

    /**
     * Se despliega sola al caerse el enlace (hay que poder salir). Mientras
     * se reconecta, no: el servicio lo está arreglando solo y el mando sigue
     * en la mano. Tampoco por avisos ni por cambios de modo del PC.
     */
    fun expandsOn(link: UiLink): Boolean = link is UiLink.Disconnected
}

/**
 * Los dos textos de la cabecera plegable del GamePad (Wii U, Pro y Switch),
 * ya traducidos por quien llama. Sin Compose: se prueban en la JVM.
 */
object GamePadHeaderText {
    /**
     * Lo que dice la pastilla: con el enlace vivo, el modo que se quiere
     * («Wii U», «Switch»), que es lo que se está jugando aunque el eco del
     * receptor aún no haya llegado; si no, el estado del enlace.
     *
     * Reconectando no es asunto suyo: la pastilla pinta entonces el punto
     * latiendo ([ReconnectingLabel]) y no este texto, así que aquí ese caso
     * se reduce a «sin conexión».
     */
    fun handle(connected: Boolean, connecting: Boolean, mode: String, connectingText: String, disconnectedText: String): String = when {
        connected -> mode
        connecting -> connectingText
        else -> disconnectedText
    }

    /**
     * La línea de estado de la tarjeta, la de siempre: «J1 · GamePad» con el
     * ida y vuelta cuando se sabe («J1 · GamePad · 23 ms»). Sin el modo
     * confirmado todavía, el aviso de que se está activando.
     *
     * En la tarjeta hay sitio de sobra, así que aquí no se recorta nada por
     * ancho (la cabecera de una línea sí lo hacía).
     */
    fun status(operative: Boolean, player: Int, padName: String, rttMs: Float?, activating: String): String {
        if (!operative) return activating
        return buildString {
            append("J")
            append(player)
            append(" · ")
            append(padName)
            if (rttMs != null) append(" · ${"%.0f".format(rttMs)} ms")
        }
    }
}

/**
 * ¿Hay ahora mismo un lector de pantalla explorando por toque (TalkBack)? Se
 * mira en vivo, no una vez al entrar: encenderlo con el mando ya abierto
 * desactiva el pliegue automático en el acto, y apagarlo lo vuelve a activar
 * (iOS lee `UIAccessibility.isVoiceOverRunning` cada vez, igual).
 */
@Composable
fun rememberScreenReader(): Boolean {
    val context = LocalContext.current
    var on by remember { mutableStateOf(false) }
    DisposableEffect(context) {
        val manager = context.getSystemService(Context.ACCESSIBILITY_SERVICE) as? AccessibilityManager
        on = manager != null && manager.isEnabled && manager.isTouchExplorationEnabled
        val listener = AccessibilityManager.TouchExplorationStateChangeListener { exploring ->
            on = exploring && manager?.isEnabled == true
        }
        manager?.addTouchExplorationStateChangeListener(listener)
        onDispose { manager?.removeTouchExplorationStateChangeListener(listener) }
    }
    return on
}

/**
 * El mecanismo de la cabecera plegable, sin decidir qué va dentro: la
 * pastilla ([label], y a su lado lo que pida [beside], que se ve siempre) y,
 * debajo, la tarjeta que dibuje [card] con el marco de siempre.
 *
 * Nace desplegada y se pliega sola a los cuatro segundos
 * ([HeaderCollapse.AUTO_MS]), para que la tarjeta no quede encima de los
 * botones que hay debajo. Cualquier toque dentro de la tarjeta reinicia la
 * cuenta sin robarle el toque a nadie ([PointerEventPass.Initial], sin
 * consumir). [card] recibe el «pliégate» por si algo de dentro (el teclado)
 * tiene que quitar la tarjeta de en medio.
 *
 * [topPadding]: el hueco sobre la caja táctil de 44 dp de la pastilla. Los
 * mandos de lado le dan 6 dp; el GamePad, 0, que así sus 44 dp son justo la
 * banda de cabecera y el hueco que ya tenía el layout.
 */
@Composable
fun CollapsibleHeader(
    link: UiLink,
    label: String,
    modifier: Modifier = Modifier,
    topPadding: Dp = 6.dp,
    beside: @Composable RowScope.() -> Unit = {},
    onExpandedChange: (Boolean) -> Unit = {},
    // Además de las reglas de HeaderCollapse: el GamePad no se pliega solo
    // mientras espera el eco del modo (el mando está inerte y la tarjeta es
    // lo único que explica por qué, con los chips y «Salir» a mano)
    autoCollapse: Boolean = true,
    card: @Composable (collapse: () -> Unit) -> Unit
) {
    var expanded by rememberSaveable { mutableStateOf(true) }
    // Cada toque dentro de la tarjeta sube el contador y con él vuelve a
    // empezar la cuenta atrás del pliegue
    var touches by remember { mutableIntStateOf(0) }

    val screenReader = rememberScreenReader()
    val auto = autoCollapse && HeaderCollapse.autoCollapses(link is UiLink.Connected, screenReader)
    LaunchedEffect(expanded, touches, auto) {
        if (expanded && auto) {
            delay(HeaderCollapse.AUTO_MS)
            expanded = false
        }
    }
    // Se cayó el enlace: la cabecera vuelve sola (hay que poder salir)
    val wantsExpanded = HeaderCollapse.expandsOn(link)
    LaunchedEffect(wantsExpanded) {
        if (wantsExpanded) expanded = true
    }
    val notify by rememberUpdatedState(onExpandedChange)
    LaunchedEffect(expanded) { notify(expanded) }

    Column(
        modifier = modifier.padding(top = topPadding),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            HeaderHandle(link, label, expanded) { expanded = !expanded }
            beside()
        }
        AnimatedVisibility(
            visible = expanded,
            enter = fadeIn(tween(140)) + expandVertically(tween(160)),
            exit = fadeOut(tween(120)) + shrinkVertically(tween(140))
        ) {
            HeaderCardFrame(onTouch = { touches++ }) {
                card { expanded = false }
            }
        }
    }
}

/**
 * Cabecera plegable de los mandos de Wii apaisados (mando + Nunchuk y mando
 * de lado): la pastilla con el modo y, dentro de la tarjeta, el nombre del
 * PC, los chips, el Nunchuk, el selector de mando y «Salir».
 *
 * [alwaysNunchukChip]: el chip «Nunchuk» siempre (el mando + Nunchuk, donde
 * es la forma de volver al mando de siempre); si no, con las reglas de
 * [showNunchukChip]. [onKeyboard]: abre el teclado de Cemu (y pliega); solo
 * sale en modo Wii U. [onExpandedChange] avisa a la pantalla (el mando +
 * Nunchuk esconde su tarjeta de «¿está bien así?» mientras está desplegada).
 */
@Composable
fun CollapsibleHeader(
    link: UiLink,
    showChips: Boolean,
    modifier: Modifier = Modifier,
    alwaysNunchukChip: Boolean = false,
    onKeyboard: (() -> Unit)? = null,
    onExpandedChange: (Boolean) -> Unit = {},
    onDisconnect: () -> Unit
) {
    CollapsibleHeader(
        link = link,
        label = handleLabel(link),
        modifier = modifier,
        onExpandedChange = onExpandedChange
    ) { collapse ->
        HeaderCard(
            link = link,
            showChips = showChips,
            alwaysNunchukChip = alwaysNunchukChip,
            onKeyboard = onKeyboard?.let { open -> { collapse(); open() } },
            onDisconnect = onDisconnect
        )
    }
}

/**
 * La pastilla: caja táctil de 44 dp de alto (el dedo la coge sin apuntar)
 * con una píldora visible de 26 dp dentro, el modo y la flecha.
 */
@Composable
private fun HeaderHandle(link: UiLink, label: String, expanded: Boolean, onToggle: () -> Unit) {
    val state = stringResource(if (expanded) R.string.header_hide else R.string.header_show)
    // Plegada, la pastilla es lo único que se ve del enlace: el lector tiene
    // que decir también el modo («Controles · Dolphin · Nunchuk»), que si no
    // el modo y el estado del enlace se quedan fuera del lector y de la voz
    val name = stringResource(R.string.header_controls) + " · " + handleSpoken(link, label)
    Box(
        modifier = Modifier
            .height(44.dp)
            .clickable(role = Role.Button, onClick = onToggle)
            .semantics(mergeDescendants = true) {
                contentDescription = name
                stateDescription = state
            },
        contentAlignment = Alignment.Center
    ) {
        Row(
            modifier = Modifier
                .height(26.dp)
                // ancho = el del texto + 44, entre 96 y 240 dp
                .widthIn(min = 96.dp, max = 240.dp)
                .background(PepoColors.Card.copy(alpha = 0.85f), RoundedCornerShape(13.dp))
                .padding(horizontal = 22.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterHorizontally)
        ) {
            val style = MaterialTheme.typography.bodyMedium.copy(color = PepoColors.TextDim)
            if (link is UiLink.Reconnecting) {
                ReconnectingLabel(link, style, Modifier.weight(1f, fill = false))
            } else {
                Text(
                    label,
                    style = style,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f, fill = false)
                )
            }
            Text(if (expanded) "▴" else "▾", style = style)
        }
    }
}

/**
 * Lo que se lee de la pastilla: lo mismo que se ve. Reconectando, el texto
 * del punto latiendo ([ReconnectingLabel]), que no es el de la pastilla.
 */
@Composable
private fun handleSpoken(link: UiLink, label: String): String = when (link) {
    is UiLink.Reconnecting -> stringResource(R.string.status_reconnecting, link.pcName)
    else -> label
}

/** Lo que dice la pastilla de los mandos de Wii: el modo, como siempre. */
@Composable
private fun handleLabel(link: UiLink): String = when {
    link !is UiLink.Connected -> stringResource(
        if (link is UiLink.Connecting) R.string.status_connecting else R.string.status_disconnected
    )
    // el mando lleva su propio Nunchuk (esta pantalla solo sale así)
    link.mode == LinkState.MODE_DOLPHIN && link.ownNunchuk -> stringResource(R.string.mode_dolphin_nunchuk)
    isWiiUAsWiimote(link) -> stringResource(R.string.wiiu_as_wiimote)
    else -> modeLabel(link.mode)
}

/**
 * El marco de la tarjeta, igual para todas las cabeceras plegables: bajo la
 * pastilla, centrada, con tope de ancho, y con los dedos vigilados.
 */
@Composable
private fun HeaderCardFrame(onTouch: () -> Unit, content: @Composable ColumnScope.() -> Unit) {
    val shape = RoundedCornerShape(14.dp)
    Column(
        modifier = Modifier
            .padding(top = 4.dp)
            .widthIn(max = 420.dp)
            .background(PepoColors.Card, shape)
            .border(1.5.dp, PepoColors.CardBorder, shape)
            // mirar los dedos antes que nadie sin quedárselos: la cuenta atrás
            // del pliegue vuelve a empezar y el chip de dentro recibe su toque
            .pointerInput(Unit) {
                awaitPointerEventScope {
                    while (true) {
                        val event = awaitPointerEvent(PointerEventPass.Initial)
                        if (event.changes.any { it.pressed != it.previousPressed }) onTouch()
                    }
                }
            }
            // ...y quedarse el dedo que no coja ningún hijo: la tarjeta tapa a
            // B, Z y C, y sin esto «Pulsar deslizando» los pulsaría a través
            .pressShield()
            .padding(horizontal = 14.dp, vertical = 8.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(6.dp),
        content = content
    )
}

/**
 * La tarjeta de los mandos de Wii: las mismas filas que tenía la cabecera de
 * siempre, una debajo de otra ahora que no hay que meterlas en una línea.
 */
@Composable
private fun HeaderCard(
    link: UiLink,
    showChips: Boolean,
    alwaysNunchukChip: Boolean,
    onKeyboard: (() -> Unit)?,
    onDisconnect: () -> Unit
) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(20.dp)
    ) {
        if (link is UiLink.Reconnecting) {
            ReconnectingLabel(link, MaterialTheme.typography.bodyMedium)
        } else {
            Text(
                when (link) {
                    is UiLink.Connected -> link.pcName
                    is UiLink.Connecting -> stringResource(R.string.status_connecting)
                    else -> stringResource(R.string.status_disconnected)
                },
                style = MaterialTheme.typography.bodyMedium,
                // un nombre de PC largo no echa a «Salir» fuera de la tarjeta
                modifier = Modifier.widthIn(max = 160.dp),
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
        }
        // Modo Wii U: texto para el teclado en pantalla de Cemu
        if (onKeyboard != null && link is UiLink.Connected && link.mode == LinkState.MODE_CEMU) {
            KeyboardButton(compact = true, onClick = onKeyboard)
        }
        TextButton(onClick = onDisconnect) {
            Text(stringResource(R.string.exit), color = PepoColors.Error, style = MaterialTheme.typography.bodyMedium)
        }
    }
    if (link is UiLink.Connected) {
        if (showModeChips(link, showChips)) {
            ModeChips(
                current = link.mode,
                supportsCemu = link.supportsCemu,
                supportsSwitch = link.supportsSwitch,
                androidReceiver = link.platform == "android",
                compact = true
            )
        }
        if (alwaysNunchukChip || showNunchukChip(link)) {
            NunchukChip(link, compact = true)
        }
        if (isWiiUAsWiimote(link)) {
            PadSelector(link, compact = true)
        }
    }
}

File: ios/PepoMote/UI/Components/CollapsibleHeader.swift
import SwiftUI
import UIKit

/// Cuándo se pliega y cuándo se abre sola la cabecera de los mandos apaisados,
/// y cuánto mide su pastilla. Lógica pura (misma que Android
/// `HeaderCollapse.kt`), para poder probarla sin pintar nada.
enum HeaderCollapse {
    /// Lo que tarda en plegarse sola: 4 s, los mismos 4000 ms de Android.
    /// `TimeInterval` son SEGUNDOS (por eso no se llama `autoMs`: pasárselo a
    /// una API de milisegundos plegaría la cabecera a los 4000 s). Cualquier
    /// toque dentro de la tarjeta vuelve a empezar la cuenta.
    static let autoSeconds: TimeInterval = 4.0

    /// Ancho máximo de la tarjeta desplegada.
    static let cardWidth: CGFloat = 420

    /// Caja táctil de la pastilla (44 de alto, como cualquier botón).
    static let handleHeight: CGFloat = 44

    /// Píldora que se ve dentro de esa caja.
    static let pillHeight: CGFloat = 26

    /// ¿Se pliega sola? Solo con el mando en juego (hay PC al otro lado) y sin
    /// lector de pantalla: con VoiceOver la tarjeta se queda hasta que la
    /// cierren, que plegarse sola se lleva el foco.
    static func autoCollapses(connected: Bool, screenReader: Bool) -> Bool {
        connected && !screenReader
    }

    /// ¿Se abre sola al cambiar el enlace? Solo al quedarse sin PC, que es
    /// cuando hacen falta «Salir» y los chips; reconectando no, que el mando
    /// sigue en la mano y la tarjeta taparía los botones.
    static func expandsOn(_ link: UiLink) -> Bool {
        if case .disconnected = link { return true }
        return false
    }

    /// Ancho de la caja táctil de la pastilla: el texto con su aire (22 a cada
    /// lado), nunca menos de 96 ni más de 240.
    static func handleWidth(textWidth: CGFloat) -> CGFloat {
        Swift.min(Swift.max(96, textWidth + 44), 240)
    }
}

/// Los dos textos de la cabecera plegable del GamePad (Wii U, Pro y Switch),
/// ya traducidos por quien llama. Sin SwiftUI: se prueban solos, con los
/// mismos vectores que Android (`GamePadHeaderText` / `GamePadHeaderTextTest`).
enum GamePadHeaderText {
    /// Lo que dice la pastilla: con el enlace vivo, el modo que se quiere
    /// («Wii U», «Switch»), que es lo que se está jugando aunque el eco del
    /// receptor aún no haya llegado; si no, el estado del enlace.
    static func handle(connected: Bool, connecting: Bool, mode: String, connectingText: String, disconnectedText: String) -> String {
        if connected { return mode }
        return connecting ? connectingText : disconnectedText
    }

    /// La línea de estado de la tarjeta, la de siempre: «J1 · GamePad» con el
    /// ida y vuelta cuando se sabe («J1 · GamePad · 23 ms»). Sin el modo
    /// confirmado todavía, el aviso de que se está activando.
    ///
    /// En la tarjeta hay sitio de sobra, así que aquí no se recorta nada por
    /// ancho (la cabecera de una línea sí lo hacía).
    static func status(operative: Bool, player: Int, padName: String, rttMs: Float?, activating: String) -> String {
        if !operative { return activating }
        var s = "J\(player) · \(padName)"
        if let rtt = rttMs { s += " · \(String(format: "%.0f", rtt)) ms" }
        return s
    }
}

/// Plegar la tarjeta desde dentro: abrir el teclado la pliega (tapa la
/// pantalla, y al volver el mando queda despejado).
struct HeaderCollapseAction {
    let collapse: () -> Void
}

private struct HeaderCollapseKey: EnvironmentKey {
    static let defaultValue = HeaderCollapseAction(collapse: {})
}

extension EnvironmentValues {
    /// Solo vale dentro de la tarjeta de `CollapsibleHeader`; fuera no hace nada.
    var headerCollapse: HeaderCollapseAction {
        get { self[HeaderCollapseKey.self] }
        set { self[HeaderCollapseKey.self] = newValue }
    }
}

/// Cabecera plegable de los mandos apaisados: una pastilla centrada arriba con
/// el modo y, desplegada bajo ella, la tarjeta con el nombre del PC, los chips
/// y «Salir». Entra abierta, se pliega sola a los 4 s y se vuelve a abrir
/// tocando la pastilla, para que nada quede encima de B/Z/C mientras se juega.
///
/// Al lado de la pastilla puede ir algo que NUNCA se pliega (`beside`): el
/// GamePad pone ahí «Teclado», que jugando tiene que estar a un toque.
///
/// Va la última de la pantalla (`.overlay(alignment: .top)`): la tarjeta gana
/// a los botones que tape.
struct CollapsibleHeader<Panel: View, Beside: View>: View {
    /// Texto de la pastilla (el modo, el mismo de la cabecera de siempre).
    private let handleLabel: String
    /// Hay PC al otro lado: sin él la tarjeta no se pliega sola.
    private let connected: Bool
    /// Aire sobre la pastilla. El GamePad lo pone a 0: así la caja táctil (44)
    /// ocupa justo la banda de la cabecera y su hueco (38 + 6) y nada de lo de
    /// abajo se mueve.
    private let topPadding: CGFloat
    /// Para que la pantalla sepa si la tarjeta tapa algo (la pregunta del lado).
    private let onExpandedChange: ((Bool) -> Void)?
    /// Lo que va junto a la pastilla, siempre a la vista.
    private let beside: () -> Beside
    private let panel: () -> Panel

    @ObservedObject private var link = LinkState.shared
    /// Se entra con la tarjeta abierta (ahí están los chips y «Salir»).
    @State private var expanded = true
    /// Cuenta atrás del pliegue automático.
    @State private var collapse: DispatchWorkItem?

    init(
        handleLabel: String,
        connected: Bool,
        topPadding: CGFloat = 6,
        onExpandedChange: ((Bool) -> Void)? = nil,
        @ViewBuilder beside: @escaping () -> Beside,
        @ViewBuilder panel: @escaping () -> Panel
    ) {
        self.handleLabel = handleLabel
        self.connected = connected
        self.topPadding = topPadding
        self.onExpandedChange = onExpandedChange
        self.beside = beside
        self.panel = panel
    }

    var body: some View {
        VStack(spacing: 4) {
            // Pastilla y lo de al lado, centrados como un bloque
            HStack(spacing: 8) {
                handle
                beside()
            }
            // Sin aire arriba, la fila entera (también lo de al lado, que
            // crece con la letra grande) se queda en la banda de 44
            .frame(height: topPadding == 0 ? HeaderCollapse.handleHeight : nil)
            if expanded { card }
        }
        .padding(.top, topPadding)
        .frame(maxWidth: .infinity, alignment: .top)
        .onAppear {
            // La pantalla se entera del estado inicial: si la cabecera se
            // quita y vuelve (pantalla completa del GamePad), nace abierta
            onExpandedChange?(expanded)
            poke()
        }
        // Al salir de la pantalla no se queda ninguna cuenta atrás viva
        .onDisappear { cancelCollapse() }
        .onChange(of: connected) { _ in poke() }
        .onChange(of: link.link) { l in
            if HeaderCollapse.expandsOn(l) { setExpanded(true) }
        }
    }

    // MARK: pastilla

    private var handle: some View {
        let text = handleLabel + (expanded ? " ▴" : " ▾")
        let pill = Self.pillHeight()
        return Button(action: { setExpanded(!expanded) }) {
            Text(text)
                .pepoBody()
                .lineLimit(1)
                .truncationMode(.tail)
                .padding(.horizontal, 10)
                .frame(width: HeaderCollapse.handleWidth(textWidth: Self.textWidth(text)), height: pill)
                .background(Pepo.card.opacity(0.85))
                .clipShape(RoundedRectangle(cornerRadius: 13, style: .continuous))
                // Sin aire arriba (el GamePad) la caja táctil no puede crecer
                // con la letra grande: 44 es justo la banda más su hueco, y de
                // ahí para abajo empieza el selector de mando
                .frame(height: topPadding == 0 ? HeaderCollapse.handleHeight : Swift.max(HeaderCollapse.handleHeight, pill))
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        // Plegada, la pastilla es lo único que se ve del enlace: el lector tiene
        // que decir también el modo («Controles · Dolphin · Nunchuk»)
        .accessibilityLabel(tr("header_controls") + " · " + handleLabel)
        .accessibilityValue(expanded ? tr("header_hide") : tr("header_show"))
        .accessibilityAddTraits(.isButton)
    }

    /// La tipografía de la pastilla (`pepoBody`, Nunito de 14) ya escalada por
    /// Dynamic Type: `Font.custom` crece con la letra del sistema, así que
    /// medir con los 14 pt pelados truncaba el modo con letra grande.
    private static func pillFont() -> UIFont {
        let base = UIFont(name: "Nunito-Regular", size: 14) ?? UIFont.systemFont(ofSize: 14)
        return UIFontMetrics.default.scaledFont(for: base)
    }

    /// Lo que ocupa el texto con la tipografía de la pastilla (bodyMedium).
    private static func textWidth(_ text: String) -> CGFloat {
        (text as NSString).size(withAttributes: [.font: pillFont()]).width.rounded(.up)
    }

    /// Alto de la píldora: los 26 de siempre, o lo que pida la letra grande.
    private static func pillHeight() -> CGFloat {
        Swift.max(HeaderCollapse.pillHeight, (pillFont().lineHeight + 6).rounded(.up))
    }

    // MARK: tarjeta

    private var card: some View {
        panel()
            .environment(\.headerCollapse, HeaderCollapseAction(collapse: { setExpanded(false) }))
            .padding(.horizontal, 14)
            .padding(.vertical, 10)
            // Como el `widthIn(max = 420)` de Compose: la tarjeta abraza su
            // contenido y solo topa en 420. Sin el `fixedSize`, un marco
            // flexible de SwiftUI se come todo lo que le ofrecen y la tarjeta
            // medía SIEMPRE 420, tapando mucho más de B/Z/C y de 1/2
            .frame(maxWidth: HeaderCollapse.cardWidth)
            .fixedSize(horizontal: true, vertical: false)
            .pepoCard(radius: 14)
            // Cualquier toque dentro reinicia la cuenta atrás, sin robarles el
            // toque a los chips ni a «Salir»
            .simultaneousGesture(DragGesture(minimumDistance: 0).onChanged { _ in poke() })
            .transition(.opacity.combined(with: .move(edge: .top)))
    }

    // MARK: abrir, plegar y la cuenta atrás

    private func setExpanded(_ on: Bool) {
        if on != expanded {
            withAnimation(.easeOut(duration: 0.16)) { expanded = on }
            onExpandedChange?(on)
        }
        if on { scheduleCollapse() } else { cancelCollapse() }
    }

    private func scheduleCollapse() {
        cancelCollapse()
        guard HeaderCollapse.autoCollapses(connected: connected, screenReader: UIAccessibility.isVoiceOverRunning) else { return }
        let work = DispatchWorkItem { setExpanded(false) }
        collapse = work
        DispatchQueue.main.asyncAfter(deadline: .now() + HeaderCollapse.autoSeconds, execute: work)
    }

    private func cancelCollapse() {
        collapse?.cancel()
        collapse = nil
    }

    /// Un toque dentro de la tarjeta (o entrar en la pantalla): la cuenta atrás
    /// vuelve a empezar.
    private func poke() {
        guard expanded else { return }
        scheduleCollapse()
    }
}

/// Cabecera sin nada al lado de la pastilla: los dos mandos de Wii apaisados,
/// que la llaman igual que siempre.
extension CollapsibleHeader where Beside == EmptyView {
    init(
        handleLabel: String,
        connected: Bool,
        topPadding: CGFloat = 6,
        onExpandedChange: ((Bool) -> Void)? = nil,
        @ViewBuilder panel: @escaping () -> Panel
    ) {
        self.init(
            handleLabel: handleLabel,
            connected: connected,
            topPadding: topPadding,
            onExpandedChange: onExpandedChange,
            beside: { EmptyView() },
            panel: panel
        )
    }
}
